CORE-1412 EC2 Tags & AutoScaling Discovery Providers

athompson-r7 opened the pull request:
This is a ruby gem that provides functionality to list all of the members of a particular group of nodes.

Per the README usage looks something like this:

``` ruby
instances = Cluster::Discovery.discover(
  'ec2_tag',
  aws_region: 'us-east-1',
  aws_tags: [{ key: 'Service', value: 'router' }])

instances.map(&:instance_id)
```

athompson-r7 committed:
CORE-1412 Start of the discovery api

This commit provides a proxy method to access discover methods of
different discovery providers.

athompson-r7 committed:
CORE-1412 Initial ec2_tag implemtation

This commit adds the start of the ec2_tag provider.

athompson-r7 committed:
CORE-1412 Style and doc fixes

athompson-r7 committed:
CORE-1412 Extract ec2 client

This commit extract ec2 client into base client

athompson-r7 committed:
CORE-1412 Add AutoScaling provider

This commit adds an auto scaling provider.  This reuses the tag provider
to perform tag searches.

athompson-r7 committed:
Revert "CORE-1412 Extract ec2 client"

This reverts commit 6897417f421f1b5ab230962c517173636af7eeb4.

athompson-r7 committed:
CORE-1412 Fix for invalid key names in tag search

athompson-r7 committed:
CORE-1412 Update readme

athompson-r7 committed:
CORE-1412 Fix class name

athompson-r7 committed:
CORE-1412 Fixes proxy methods

This commit fixes the proxy methods so that they work(TM).  This also
updates the docs so that they are inline with the code.

fmitchell-r7 commented on lib/cluster/discovery/ec2/tag.rb:
```diff
@@ -0,0 +1,65 @@
+            a << page.reservations.map(&:instances)
+          end
+          instances = instances.flatten.compact
+          instances.sort_by! { |x| "#{x.launch_time.to_i}-#{x.instance_id}" }
```
I kind of feel like sorting the discovered instances is one of those things that should be left up to the user.

athompson-r7 replied:
It seemed sensible to return a sorted list to the user to try to maintain some order.

For very large sets this may not be sensible.  I'll rip it out.

athompson-r7 committed:
CORE-1412 Remove sort from ec2 tags discovery

This commit removes the sort from the ec2_tags provider.  This should be
handled by the user.

fmitchell-r7 commented:
:+1: Looks good to me.
